driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2288,230 @@ public void run() {
                 }
             }
         }
+
+        private class SchemaRefreshRequest {
+
+            private final String keyspace;
+
+            private final String table;
+
+            SchemaRefreshRequest(String keyspace, String table) {
+                this.keyspace = keyspace;
+                this.table = table;
+            }
+
+            /**
+             * Coalesce schema refresh requests.
+             * The algorithm is simple: if more than 2 keyspaces
+             * need refresh, then refresh the entire schema;
+             * otherwise if more than 2 tables in the same keyspace
+             * need refresh, then refresh the entire keyspace.
+             *
+             * @param that the other request to merge with the current one.
+             * @return A coalesced request
+             */
+            SchemaRefreshRequest coalesce(SchemaRefreshRequest that) {
+                if(this.keyspace == null || that.keyspace == null)
+                    return new SchemaRefreshRequest(null, null);
+                if(!this.keyspace.equals(that.keyspace))
```

Comment:
One small concern I have here is that if you have changes to two keyspaces in this interval refreshing the entire schema could be pretty costly, as chances are if you have two keyspaces altered within a small interval that you have a pretty complex schema.   The same logic can be applied to tables within the same keyspace, although I think the way this currently works is reasonable.  We could have some kind of threshold where we decide if X keyspace/tables have changed only then refresh the entire schema/keyspace, otherwise refresh each individually, though don't know how important it is to have something like that.